pkgs/native_assets_cli/lib/src/model/link_config.dart

```diff
@@ -0,0 +1,168 @@
+part of '../api/link_config.dart';
+
+/// The input to the linking script.
+///
+/// It consists of the [_buildConfig] already passed to the build script, the
+/// [assets] from the build step, and the [resources] generated during the
+/// kernel compilation.
+class LinkConfigImpl extends HookConfigImpl implements LinkConfig {
+  @override
+  final List<AssetImpl> assets;
+
+  final BuildConfigImpl _buildConfig;
+
+  @override
+  final List<Resource> resources;
+
+  final _LinkConfigArgs _args;
+
+  LinkConfigImpl._(
+    this._args, {
+    required this.assets,
+    required BuildConfigImpl buildConfig,
+    required ResourceIdentifiers? resourceIdentifiers,
+  })  : _buildConfig = buildConfig,
+        resources = fromIdentifiers(resourceIdentifiers);
+
+  @override
+  Uri get configFile => outputDirectory.resolve('../link_config.json');
+
+  @override
+  Uri get outputDirectory => _buildConfig.outputDirectory;
+
+  @override
+  String get outputName => 'link_output.json';
+
+  @override
+  String get packageName => _buildConfig.packageName;
+
+  @override
+  Uri get packageRoot => _buildConfig.packageRoot;
+
+  @override
+  Uri get script => packageRoot.resolve('hook/').resolve(Hook.link.scriptName);
+
+  @override
+  String toJsonString() =>
+      const JsonEncoder.withIndent('  ').convert(_args.toJson());
+
+  @override
+  Version get version => _buildConfig.version;
+
+  static LinkConfig fromArguments(List<String> arguments) =>
+      _LinkConfigArgs.fromArguments(arguments).toLinkConfig();
+
+  factory LinkConfigImpl.fromValues({
+    required Uri? resourceIdentifierUri,
+    required BuildConfigImpl buildConfig,
+    required List<AssetImpl> assetsForLinking,
+  }) =>
+      _LinkConfigArgs(
+        assetsForLinking: assetsForLinking,
+        buildConfig: buildConfig,
+        resourceIdentifierUri: resourceIdentifierUri,
+      ).toLinkConfig();
+
+  @override
+  BuildMode get buildMode => _buildConfig.buildMode;
+
+  @override
+  CCompilerConfig get cCompiler => _buildConfig.cCompiler;
+
+  @override
+  bool get dryRun => _buildConfig.dryRun;
+
+  @override
+  Iterable<String> get supportedAssetTypes => _buildConfig.supportedAssetTypes;
+
+  @override
+  int? get targetAndroidNdkApi => _buildConfig.targetAndroidNdkApi;
+
+  @override
+  Architecture? get targetArchitecture => _buildConfig.targetArchitecture;
+
+  @override
+  IOSSdk get targetIOSSdk => _buildConfig.targetIOSSdk;
+
+  @override
+  OS get targetOS => _buildConfig.targetOS;
+
+  /// The version of [BuildConfigImpl].
+  ///
+  /// This class is used in the protocol between the Dart and Flutter SDKs
+  /// and packages through build hook invocations.
+  ///
+  /// If we ever were to make breaking changes, it would be useful to give
+  /// proper error messages rather than just fail to parse the JSON
+  /// representation in the protocol.
+  static Version latestVersion = Version(1, 0, 0);
+}
+
+List<Resource> fromIdentifiers(ResourceIdentifiers? resourceIdentifiers) =>
```

Comment:
Why is  `class Resource` not an abstract interface and implemented by `class Identifier`? That's the pattern we use for the `BuildConfig` and it's parts. It would make our code more internally consistent to follow this pattern. (Instead of constructing new objects in this method.)

Is there anything preventing us from continuing that existing pattern?